File: app/src/main/java/com/szymkowski/personaltrainercompanion/core/DbCollaborator.java
package com.szymkowski.personaltrainercompanion.core;

import android.database.sqlite.SQLiteDatabase;
import android.util.Log;

import com.j256.ormlite.support.ConnectionSource;
import com.j256.ormlite.table.TableUtils;

import java.lang.reflect.ParameterizedType;
import java.sql.SQLException;

public abstract class DbCollaborator<T> {

    private final Class klazz;

    public DbCollaborator() {
        this.klazz  = (Class) ((ParameterizedType) getClass()
                .getGenericSuperclass()).getActualTypeArguments()[0];
    }

    public void createTableIfNotExists(ConnectionSource connectionSource) throws SQLException {
        Log.i(DbCollaborator.this.getClass().getName(), "onCreate");
        TableUtils.createTableIfNotExists(connectionSource, klazz);
    }


    public void upgradeTable(SQLiteDatabase db, ConnectionSource connectionSource, int oldVersion, int newVersion) throws SQLException {
        throw new RuntimeException("not implemented!");
    }

}

File: app/src/main/java/com/szymkowski/personaltrainercompanion/core/DbCollaborator.kt
package com.szymkowski.personaltrainercompanion.core

import android.database.sqlite.SQLiteDatabase
import android.util.Log

import com.j256.ormlite.support.ConnectionSource
import com.j256.ormlite.table.TableUtils

import java.lang.reflect.ParameterizedType
import java.sql.SQLException

/*
This whole byzantine construct only serves to prevent domain entity objects from flying around the whole application.
I want them only available from their packages.
 */

abstract class DbCollaborator<T> protected constructor() {

    private val klazz: Class<Any>

    init {
        this.klazz = (javaClass.genericSuperclass as ParameterizedType).actualTypeArguments[0] as Class<Any>
    }

    @Throws(SQLException::class)
    fun createTableIfNotExists(connectionSource: ConnectionSource) {
        Log.i(this@DbCollaborator.javaClass.name, "onCreate")
        TableUtils.createTableIfNotExists(connectionSource, klazz)
    }


    @SuppressWarnings("EmptyMethod", "RedundantThrows", "UnusedParameters")
    @Throws(SQLException::class)
    abstract fun upgradeTable(db: SQLiteDatabase, connectionSource: ConnectionSource, oldVersion: Int, newVersion: Int)

}
